!!!FEATURE: Lock the UI and present a modal during publish/discard

grebaldi opened the pull request:
<div align=center>

<table>
<tbody>
<tr>
<td><img src="https://github.com/neos/neos-ui/assets/2522299/565a5490-3598-4217-9878-6742ab8e6817" alt=""/></td>
<td><img src="https://github.com/neos/neos-ui/assets/2522299/027c1c30-e740-4d33-813c-7d0423c86cd4" alt=""/></td>
</tr>
</tbody>
</table>
</div>

solves: #3743
builds on:  #3752

The PublishDiscardDialog is a modal that is supposed to pop up whenever an editor starts publishing or discarding their workspace.

Editors need to confirm their intent, after which they are presented with a process indicator that blocks the screen as long as the remote operation is not finished yet.

Finally, Editors see a success message or an error message - depending on the outcome. In both cases, they need to acknowledge the info. If they see an error message, they are also given an opportunity to retry the action.

**Publish Workflow**

<details>
<summary>Show Video</summary>

https://github.com/neos/neos-ui/assets/2522299/e7e7f54a-406f-4400-b59b-96796abee64a

</details>

For publishing, users are now first presented with a confirmation dialog. Formerly, publishing didn't show this dialog. This PR makes both `Publish` and `Discard` consistent in this regard (as you wouldn't want to accidentally publish anything either). 

If the user confirms the action, they see an uncancellable modal that illustrates the ongoing operation. Formerly, the operation was only indicated by a spinning icon alongside the `PublishingDropdown` head. The UI remained usable during the operation, which could potentially lead to invalid actions. Now, the UI is locked, including a `beforeunload` handler that warns the user if they're about to reload or leave the page.

The progress indicator remains open for as long as any async operation is still running. This includes the publishing operation itself, but also the adjacent reload of nodes and the reload of the guest frame. Only after the UI has reached a stable and consistent state, the lock is released.

Finally, the user is presented with a dialog that lets them acknowledge the operation. Formerly, a successful publishing process would have been indicated by a fleeting flash message. The acknowledgement dialog remains open until the user has taken notice. This way users can safely leave the tab, do something else, and return without losing the context of what their last action has been.

**Discard Workflow**

<details>
<summary>Show Video</summary>

https://github.com/neos/neos-ui/assets/2522299/2d5e0453-923d-496d-a5f1-ad4e3f27b4ea

</details>

The `Discard` workflow behaves exactly like the `Publish` workflow. The only difference is in the coloring and the progress indicator illustration.

**Error Handling**


<details>
<summary>Show Video</summary>

https://github.com/neos/neos-ui/assets/2522299/1cccee8f-0382-488f-b2e1-10f9e20e1f0f

</details>

Formerly, errors during `Publish` or `Discard` would have been displayed using a persistent flash message. This PR introduces a new component that can be embedded inside a dialog and that shows more detailed error information in `FLOW_CONTEXT=Development`.

Also new: Users are offered a "Try Again" button, when `Publish` or `Discard` fails. 

**Remaining TODOs**

- [x] Create `<PublishDiscardDialog/>`
- [x] Move old I18n labels into Neos.Neos.Ui and initialize new ones
- [x] Create Follow-Up issue or PR for labels in Neos.Neos
  - https://github.com/neos/neos-development-collection/pull/4990
- [x] Wire up the PublishDiscardDialog component with redux
- [x] Split out `CR.Publishing` state from `CR.Workspaces`
  - [x] Summarize actions `abortPublish`, `confirmPublish`, `finishPublish`, `acknowledgePublish`, `abortDiscard`, `confirmDiscard`, `finishDiscard`, `acknowledgeDiscard` -> `cancel`, `confirm`, `finish`, `acknowledge`
- [x] Rename PublishDiscard -> Publishing everywhere
  - [x] Rename PublishDiscardDialog -> PublishingDialog
  - [x] Rename PublishDiscardMode -> PublishingMode
    - [x] Rename PublishDiscardMode.PUBLISHING -> PublishingMode.PUBLISH
    - [x] Rename PublishDiscardMode.DISCARDING -> PublishingMode.DISCARD
  - [x] Rename PublishDiscardScope -> PublishingScope
  - [x] Rename PublishDiscardPhase -> PublishingPhase
- [x] Merge publish and discard saga
- [x] Remove ad-hoc `NODE_HAS_BEEN_CREATED` constant and make `TypeOfChange` an enum
- [x] Better Error Handling
  - [x] Use direct responses from publish/discard endpoints rather than FeedbackCollection to gain better control of error handling
  - [x] Handle JavaScript errors more extensively in publishing saga
  - [x] Switch action buttons in error case and rename: `[Try again]` `[Okay]` -> `[Cancel]` `[Try again]`
- [x] Implement Retry
- [x] Refer to target workspace name in success message
- [x] Adjust E2E tests
- [x] Remove legacy `<DiscardDialog/>`
- [x] Remove legacy `actions.UI.Remote` actions
  - [x] Remove `startPublishing()`
  - [x] Remove `finishPublishing()`
  - [x] Remove `startDiscarding()`
  - [x] Remove `finishDiscarding()`
- [x] Add `beforeunload` handler during publish/discard
- [x] Fetch updated workspace info from server after publish/discard
- [x] **!!! Revert commit https://github.com/neos/neos-ui/pull/3759/commits/79de33b4c1ff4baf425e7b951eeed186214ab870 !!!**
- [x] Fix: All 0 change(s) in workspace "user-admin" were sucessfully discarded.

**Follow-Ups**

- https://github.com/neos/neos-ui/pull/3762

grebaldi committed:
FEATURE: Create `<PublishDiscardDialog/>`

The PublishDiscardDialog is a modal that is supposed to pop up whenever
an editor starts publishing or discarding their workspace.

Editors need to confirm their intent, after which they are presented
with a process indicator that blocks the screen as long as the remote
operation is not finished yet.

Finally, Editors see a success message or an error message - depending
on the outcome. In both cases, they need to acknowledge the info. If
they see an error message, they are also given an opportunity to retry
the action.

This commit creates the visual implementation of the dialog but does not
wire it to the application state yet.

grebaldi committed:
TASK: Add dedicated xliff translations for PublishDiscardDialog

grebaldi committed:
TASK: Wire up the PublishDiscardDialog component with redux

grebaldi committed:
TASK: Split out `CR.Publishing` state from `CR.Workspaces`

grebaldi committed:
TASK: Rename PublishDiscardDialog -> PublishingDialog

grebaldi committed:
TASK: Rename PublishDiscard* -> Publishing* everywhere

Rename PublishDiscardMode -> PublishingMode
Rename PublishDiscardMode.PUBLISHING -> PublishingMode.PUBLISH
Rename PublishDiscardMode.DISCARDING -> PublishingMode.DISCARD
Rename PublishDiscardScope -> PublishingScope
Rename PublishDiscardPhase -> PublishingPhase

grebaldi committed:
TASK: Merge publish and discard sagas

grebaldi committed:
TASK: Remove ad-hoc `NODE_HAS_BEEN_CREATED` constant and make `TypeOfChange` an enum

grebaldi committed:
TASK: Introduce package `@neos-project/neos-ui-error`

This package provides utilities to analyze errors of various types. It
provides the `<ErrorView/>` component that can be embedded in other
components to display those errors. In `FLOW_CONTEXT=Development` the
`<ErrorView/>` component will also display meta information (e.g.
exception class name, stack trace, etc.).

grebaldi committed:
TASK: Use direct responses from publish/discard endpoints rather than FeedbackCollection

grebaldi committed:
TASK: Handle errors in publishing workflow with `@neos-project/neos-ui-error`

grebaldi committed:
TASK: Switch and rename action buttons of ResultDialog in error case

grebaldi committed:
TASK: Implement Publishing Retry

grebaldi committed:
TASK: Add diagram to publishing dialog to illustrate the operation

grebaldi committed:
TASK: Fix phpstan error

grebaldi committed:
TASK: Temporarily use `publishingBonanza` for Neos.Neos 9.0 in E2E tests

This commit MUST be reverted before merging!

grebaldi committed:
TASK: Adjust E2E tests to new discarding behavior

mhsdesign commented:
Wooooooooow 🥳 the dialog looks super cool ;)

grebaldi committed:
TASK: Remove legacy `<DiscardDialog/>`

grebaldi committed:
TASK: Remove legacy `actions.UI.Remote` actions

grebaldi committed:
TASK: Add `beforeunload` handler during publish/discard

grebaldi committed:
TASK: Replace `typedKeys` shenanigans with proper explicit exports in redux-store

grebaldi committed:
TASK: Implement ReloadNodes query

grebaldi committed:
TASK: Implement reloadNodes saga using the ReloadNodes query endpoint

grebaldi committed:
TASK: Convert Publishing saga to typescript

grebaldi committed:
TASK: Use `reloadNodes` saga after discard operations

grebaldi committed:
TASK: Update WorkspaceInfo after publish/discard

grebaldi committed:
TASK: Add more precise typings to backend connector

grebaldi commented:
Unit test failures are unrelated. They are due to a broken jest snapshot that has been fixed with https://github.com/neos/neos-ui/pull/3757 for `8.3`. The fix hasn't arrived in `9.0` yet, so unit tests will be green after upmerges.

grebaldi committed:
BUGFIX: Fix wrong number of discarded changes in ResultDialog

nezaniel commented on Classes/Application/ReloadNodes/MinimalNodeForTree.php:
```diff
@@ -0,0 +1,65 @@
+    {
+    }
+
+    public static function fromNode(
```
Imho this should be named tryFromNode because it might return null, following the pattern introduced with enums

grebaldi replied:
Agreed, and adjusted :)

nezaniel commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -176,33 +184,41 @@ public function changeAction(array $changes): void
+                ->publishChangesInSite($command->siteId);
+
+            $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+            $baseWorkspaceName = $contentRepository->getWorkspaceFinder()->findOneByName($command->workspaceName)->baseWorkspaceName;
```
Good news: the Workspace model has a getCurrentBaseWorkspaceName() method for this

grebaldi replied:
Thanks for the pointer! Adjusted.

nezaniel commented on Classes/Controller/BackendServiceController.php:
```diff
@@ -212,53 +228,51 @@ public function publishChangesInSiteAction(array $command): void
+
+            $contentRepositoryId = SiteDetectionResult::fromRequest($this->request->getHttpRequest())->contentRepositoryId;
+            $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+            $baseWorkspaceName = $contentRepository->getWorkspaceFinder()->findOneByName(
```
See above

nezaniel reviewed (commented):
Love it!

I added a few nitpicks and am otherwise ready to approve

grebaldi committed:
TASK: Rename `MinimalNodeForTree::fromNode` -> `*::tryFromNode`

As suggested by @nezaniel (see:
https://github.com/neos/neos-ui/pull/3759#discussion_r1563782134)

grebaldi committed:
TASK: Use `getCurrentBaseWorkspace` of Neos.Neos' Workspace model...

...to retrieve the base workspace during publish opertaions.

As suggested by @nezaniel (see:
https://github.com/neos/neos-ui/pull/3759#discussion_r1563802249)

mhsdesign commented on Classes/Application/ReloadNodes/ReloadNodesQueryHandler.php:
```diff
@@ -0,0 +1,172 @@
+        $nodeMapBuilder->addNode($siteNode);
+
+        $gatherNodesRecursively = function (
+            &$nodeMapBuilder,
```
```suggestion
            $nodeMapBuilder,
```

mhsdesign commented on Classes/Application/ReloadNodes/ReloadNodesQueryHandler.php:
```diff
@@ -0,0 +1,172 @@
+            $query->dimensionSpacePoint,
+            VisibilityConstraints::withoutRestrictions()
+        );
+        $baseNodeTypeConstraints = NodeTypeCriteria::fromFilterString($this->baseNodeType);
```
base nodetypes doest work yet as they were not migrated correctly so they are lots of parts missing this functionality but maybe its a good idea to keep this feature considered here and we can remove it later if we want ...

mhsdesign commented on Classes/Application/ReloadNodes/ReloadNodesQueryHandler.php:
```diff
@@ -0,0 +1,172 @@
+            $ancestors,
+            $subgraph
+        ) {
+            if ($level < $this->loadingDepth || // load all nodes within loadingDepth
```
ah i see this logic is not new but moved from `NeosUiDefaultNodesOperation` ...

mhsdesign replied:
... but we dont seem to remove the NeosUiDefaultNodesOperation yet?

mhsdesign commented on Classes/Application/ReloadNodes/ReloadNodesQueryHandler.php:
```diff
@@ -0,0 +1,172 @@
+            }
+        }
+
+        /* TODO: we might use the Subtree as this may be more efficient
+         - but the logic above mirrors the old behavior better.
+        if ($loadingDepth === 0) {
+            throw new \RuntimeException('TODO: Loading Depth 0 not supported');
+        }
+        $subtree = $contentSubgraph->findSubtree([$siteNode], $loadingDepth, $nodeTypeConstraints);
+        $subtree = $subtree->getChildren()[0];
+        $this->flattenSubtreeToNodeList($nodeAccessor, $subtree, $nodeMapBuilder);*/
```
```suggestion
        /* TODO: we might use the Subtree as this may be more efficient
         - but the logic above mirrors the old behavior better.
         https://github.com/neos/neos-ui/issues/3517#issuecomment-2070274053 */
```

mhsdesign commented on packages/neos-ui-error/README.md:
```diff
@@ -0,0 +1,41 @@
+# @neos-project/neos-ui-error
```
nice idea to add some structure to our codebase :D

Should this package also contain the default error boundary? `neos-ui/src/Containers/ErrorBoundary` ... but on the other hand i guess there is little code to share and the boundary can stay were it wants

mhsdesign commented on packages/neos-ui-redux-store/src/UI/index.ts:
```diff
@@ -73,12 +47,52 @@ export interface State {
 // Export the actionTypes
 //
-export const actionTypes = typedKeys(all).reduce((acc, cur) => ({...acc, [cur]: all[cur].actionTypes}), {});
+export const actionTypes = {
```
nice cleanup that is some next level stuff otherwise 😅

mhsdesign commented on packages/neos-ui-sagas/src/CR/NodeOperations/reloadNodes.ts:
```diff
@@ -0,0 +1,128 @@
+// @ts-ignore
+import {getGuestFrameDocument} from '@neos-project/neos-ui-guest-frame/src/dom';
+
+// @TODO: This is a helper to gain type access to the available backend endpoints.
+// It shouldn't be necessary to do this, and this hack should be removed once a
+// better type API is available
+import {default as Endpoints, Routes} from '@neos-project/neos-ui-backend-connector/src/Endpoints';
+type Endpoints = ReturnType<typeof Endpoints>;
```
hmm i had to read twice till i understood :)

i would favour to export the type just there and then add the comment there as well ... here it seems to be a little misplaced and will likely never be found :O

```js
import {EndpointsType, Routes} from '@neos-project/neos-ui-backend-connector/src/Endpoints';
```

mhsdesign replied:
Will be solved as followup https://github.com/neos/neos-ui/pull/3759/files#r1575121070

mhsdesign commented on packages/neos-ui-sagas/src/Publish/index.ts:
```diff
@@ -0,0 +1,183 @@
+        const workspaceName: WorkspaceName = yield select(selectors.CR.Workspaces.personalWorkspaceNameSelector);
+        const ancestorId: NodeContextPath = yield select(ancestorIdSelector);
+
+        do {
```
well i see at least one of us understand redux 🤯 nice work will have to test each scenario as my brain will not be able to run this code in memory

mhsdesign commented on packages/neos-ui/src/Containers/Modals/PublishingDialog/ConfirmationDialog.tsx:
```diff
@@ -0,0 +1,196 @@
+
+import style from './style.module.css';
+
+const ConfirmationDialogVariants = {
```
good idea that keeps things tidy

mhsdesign reviewed (commented):
Wow nice work so much work thanks. We are lucky to have you on the team as this is some crazy state handling ;)

After reading through im not sure i understand even half of it but it looks legit ^^
I will experiment locally and - if you find the time - would like to chat with you what changed but in human language ;)

mhsdesign reviewed (approved):
Thanks for walking through this one with me and explaining everything ^^

grebaldi committed:
TASK: Clarify comment in ReloadNodesQueryHandler

Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

grebaldi committed:
TASK: Always reload nodes after publish & discard

grebaldi committed:
Merge branch 'feature/conflict-resolution-01/publishing-modal' of github.com:neos/neos-ui into feature/conflict-resolution-01/publishing-modal

mhsdesign commented on packages/neos-ui-sagas/src/Publish/index.ts:
```diff
@@ -0,0 +1,178 @@
+        }
+    };
+
+    const reloadAfterPublishing = makeReloadAfterDiscard({routes});
```
nitpick: the factory should also be renamed ;)

grebaldi replied:
Of course :D Thanks for the pointer!

grebaldi committed:
TASK: Rename factory `makeReloadAfterDiscard` -> `makeReloadAfterPublishing`

mhsdesign commented:
Okay perfect ❤️ again many thanks. Please feel free to hit the merge button. (The failures are unrelated and will be fixed by a rebase)
